drivers/serial/uart_infineon_cat1.c

```diff
@@ -0,0 +1,506 @@
+{
+	cyhal_uart_enable_event(&DEV_DATA(dev)->obj,
+				(cyhal_uart_event_t) CYHAL_UART_IRQ_RX_NOT_EMPTY,
+				UART_CAT1_INTERRUPT_PRIORITY, 1);
+}
+
+/* Disable TX interrupt */
+static void uart_cat1_irq_rx_disable(const struct device *dev)
+{
+	cyhal_uart_enable_event(&DEV_DATA(dev)->obj,
+				(cyhal_uart_event_t) CYHAL_UART_IRQ_RX_NOT_EMPTY,
+				UART_CAT1_INTERRUPT_PRIORITY, 0);
+}
+
+/* Check if UART RX buffer has a received char */
+static int uart_cat1_irq_rx_ready(const struct device *dev)
+{
+	return cyhal_uart_readable(&DEV_DATA(dev)->obj) ? 1 : 0;
+}
+
+/* Enable Error interrupts */
+static void uart_cat1_irq_err_enable(const struct device *dev)
+{
+	cyhal_uart_enable_event(&DEV_DATA(dev)->obj, (cyhal_uart_event_t)
+				(CYHAL_UART_IRQ_TX_ERROR | CYHAL_UART_IRQ_RX_ERROR),
+				UART_CAT1_INTERRUPT_PRIORITY, 1);
+}
+
+/* Disable Error interrupts */
+static void uart_cat1_irq_err_disable(const struct device *dev)
+{
+	cyhal_uart_enable_event(&DEV_DATA(dev)->obj, (cyhal_uart_event_t)
+				(CYHAL_UART_IRQ_TX_ERROR | CYHAL_UART_IRQ_RX_ERROR),
+				UART_CAT1_INTERRUPT_PRIORITY, 0);
+}
+
+/* Check if any IRQs is pending */
+static int uart_cat1_irq_is_pending(const struct device *dev)
+{
+	uint32_t intcause = Cy_SCB_GetInterruptCause(DEV_DATA(dev)->obj.base);
+
+	return (int) (intcause & (CY_SCB_TX_INTR | CY_SCB_RX_INTR));
+}
+
+/* Start processing interrupts in ISR.
+ * This function should be called the first thing in the ISR. Calling
+ * uart_irq_rx_ready(), uart_irq_tx_ready(), uart_irq_tx_complete()
+ * allowed only after this.
+ */
+static int uart_cat1_irq_update(const struct device *dev)
+{
+	int status = 1;
+
+	if (((uart_cat1_irq_is_pending(dev) & CY_SCB_RX_INTR) != 0u) &&
+	    (Cy_SCB_UART_GetNumInRxFifo(DEV_DATA(dev)->obj.base) == 0u)) {
+		status = 0;
+	}
+
+	return status;
+}
+
+static void uart_cat1_irq_callback_set(const struct device *dev,
+				       uart_irq_callback_user_data_t cb,
+				       void *cb_data)
+{
+	struct uart_cat1_data *data = DEV_DATA(dev);
+	cyhal_uart_t *uart_obj = &data->obj;
+
+	/* Store user callback info */
+	data->irq_cb = cb;
+	data->irq_cb_data = cb_data;
+
+	/* Register a uart general callback handler  */
+	cyhal_uart_register_callback(uart_obj, _uart_event_callback_irq_mode, (void *) dev);
+}
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+
+static int uart_cat1_init(const struct device *dev)
+{
+	struct uart_cat1_config *const config = DEV_CFG(dev);
+
+	cy_rslt_t result;
+	int ret;
+
+	/* Configure dt provided device signals when available */
+	ret = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Initialize the UART peripheral */
+	result = cyhal_uart_init(&DEV_DATA(dev)->obj,
+				 cat1_pinctrl_get_cyhal_gpio(config->pcfg, UART_TX_PIN),
+				 cat1_pinctrl_get_cyhal_gpio(config->pcfg, UART_RX_PIN),
+				 cat1_pinctrl_get_cyhal_gpio(config->pcfg, UART_CTS_PIN),
+				 cat1_pinctrl_get_cyhal_gpio(config->pcfg, UART_RTS_PIN),
+				 NULL, NULL);
```

Comment:
you need to use pinctrl, if your HAL takes care of pins you'll need to patch it so that it doesn't. We did the same for nRF (you can check some nRF drivers where you'll see HAL gets a special value to indicate pins are configured somewhere else)

Reply:
This is no longer allowed, you'll need to patch your HAL to not take care of pins (same as Nordic did)